### ISSUE TITLE
t.w.w.m.Model.getSubmodel caches *way* too aggressively

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11261]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)| @jyknight reported|
|-|-|
|Trac ID|trac#322|
|Type|enhancement|
|Created|2003-10-08 20:51:19Z|

<details><summary>Searchable metadata</summary>

```
trac-id__322 322
type__enhancement enhancement
reporter__jknight jknight
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ 
keywords__ 
time__1065646279000000 1065646279000000
changetime__1469694588435439 1469694588435439
version__ 
owner__ 
cc__radix cc__dp cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
When you return something from a submodelFactory in a subclass of Model, the result gets cached 
in the 'submodels' dictionary forever. Even between requests. That is very bad, as it means your 
method will only get called once *ever* per submodel-name.

fzZzy and I talked about this and believe the right thing is to eliminate the implicit caching 
entirely, and then add explicit caching *just inside one request* for people who request it.

It is not necessarily safe to ever cache a model that resulted from some random method call. Who
knows..maybe it's supposed to return a random number?

PS: my use case:
I have a model in the template '/students/%current/student_id'. I have a wmfactory_students which 
returns a subclass of Model with submodelFactory defined. Asking this class for the submodel 
'%students' uses the current logged-in user and does a DB lookup with that username. This returns 
a deferred, and will eventually return a dictionary of key/values. In this case, it's okay (good even!) 
to cache for the duration of the request, but definitely not for the entire life of the server.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="dp's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|dp commented|
|-|-|

```
#!html
<pre>
This isn't really true; there was acrobatic double-flip code in t.w.w.model to make sure that your 
MethodModel wmfactory_ methods were called every request. But the fact that acrobatic double-
flip code was required was very bad, and obviously caused bugs nobody really observed because 
they weren't using Woven the way jknight was.

The main reason things were implemented the way they were is because the woven modelChanged 
notification relies on the model identity to keep track of which views to notify (it keeps weakrefs in 
self.views). Unrelated to foom's problem, I rewrote the modelChanged notification system to use an 
external event bus for Quotient. This makes event subscription and publishing explicit, which is a 
good thing for debugging (but a somewhat bad thing for productivity). Once identity was no longer 
an issue, I was able to eliminate the submodel cache entirely (using a nasty hack) and verify that 
Quotient still worked, and that the solution fixed foom's problem.

I'm not sure when I will have time to make this into a "real" woven commit. Hopefully sometime 
within the next month or so.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
I think that it might be a good idea to be able to easily specify caching
strategies on a per-submodel basis.

"permanent": only update the model when I explicitly say to invalidate
             the cache.

"request": only cache the submodel for the duration of the request.

"none": If &lt;span model="sub"/> happens multiple times in the DOM, ask for 
        the "foo" model each time.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)|@moshez commented|
|-|-|

```
#!html
<pre>
this is starting to sound more and more like a feature request --
the code is working as intended, but the intent was misguided.
Mind downgrading?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
Some of this certainly is a feature request, and got lumped in with the actaul bug because it might 
be easier to simply rewrite the code (and add new features at the same time) than to fix the bug in 
the current code. 

I *think* the intent is to always cache for the duration of a request. That it currently caches even
between requests (sometimes) is the bug. If that is fixed so that it always clears out the cache for a 
new request, the rest is clearly additional features and can be downgraded.

Dp -- feel free to correct me, of course. ;)
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
Nope, I don't think it's ever been stated that the intent is to only
cache for the duration of the request. dp has always told me that one
of the major design decisions of woven was to make it so models only
get updated when they need to be, and are told. Of course, I think the
default should be caching for the request, and only cache longer when
specifically told :-)
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="dp's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|dp commented|
|-|-|

```
#!html
<pre>
On Oct 24, 2003, at 6:04 PM, Chris Armstrong [[Twisted issue tracker]] 
wrote:

>
> Chris Armstrong &lt;radix@...> added the comment:
>
> Nope, I don't think it's ever been stated that the intent is to only
> cache for the duration of the request.

Data generated from a wmfactory_ method is only cached for the duration 
of one request, per your desire. The implementation is baroque and 
fragile, and data which is the result of a getData call is *not* 
flushed between requests. This is what foom is observing.

> dp has always told me that one
> of the major design decisions of woven was to make it so models only
> get updated when they need to be, and are told.

This was one of the major design decisions of old woven -- partially 
because models with a given name need to have the same python identity 
even if the (possibly immutable) data changes so you can send 
notifications to the model.

> Of course, I think the
> default should be caching for the request, and only cache longer when
> specifically told :-)

But I'm reversing this design decision in nevow. Data generating 
functions are always called, every time that data is required. Fixing 
old woven to correct the exact situation foom is experiencing would be 
incredibly painful and prone to huge breakage. If someone is seriously 
attached to old woven and is having bugs, they should talk to me as 
soon as possible.

(I am planning on having an old woven api compatibility layer on top of 
nevow. microdom (Already done), Page, wvupdate, wvfactory, wmfactory, 
and wcfactory, Widget.setUp, IModel, and other specific methods. This 
backwards compatibility layer will not use the current IModel design, 
and thus will not be subject to this bug, but may be a while coming.)

dp
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)|@moshez commented|
|-|-|

```
#!html
<pre>
After consultation with radix, I'm downgrading this to "feature".
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
I would categorize this as P:bug S:wontfix, except that status doesn't exist.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
This bug is being resolved, wontfix, as woven is "almost deprecated" and unmaintained. Use nevow 
instead. See &lt;http://divmod.org/users/dp.twistd/wiki/> and the twisted-web mailing list &lt;http://
twistedmatrix.com/cgi-bin/mailman/listinfo/twisted-web>.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|GitHub commented|
|-|-|

In changeset a43775dd9af5338f5f0076b2b87f8fe7327e8174
```
#!CommitTicketReference repository="" revision="a43775dd9af5338f5f0076b2b87f8fe7327e8174"
Merge pull request #322 from twisted/oldstyle-bases-8308

Author: hawkowl
Reviewer: glyph
Fixes: #8308
```
